pkg/controllers/provisioning/scheduler.go

```diff
@@ -0,0 +1,174 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioning
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/awslabs/karpenter/pkg/utils/pod"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Controller for the resource
+type Scheduler struct {
+	kubeClient client.Client
+	controller *Controller
+}
+
+// NewScheduler constructs a controller instance
+func NewScheduler(kubeClient client.Client, controller *Controller) *Scheduler {
+	return &Scheduler{
+		kubeClient: kubeClient,
+		controller: controller,
+	}
+}
+
+// Reconcile the resource
+func (s *Scheduler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("scheduling").With("pod", req.String()))
+	pod := &v1.Pod{}
+	if err := s.kubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Ensure the pod can be provisioned, but only log
+	if err := isUnschedulable(pod); err != nil {
+		return reconcile.Result{}, nil
+	}
+	if err := validate(pod); err != nil {
+		logging.FromContext(ctx).Debugf("Ignoring pod, %s", err.Error())
+		return reconcile.Result{}, nil
+	}
+	// Schedule and requeue. If successful, will terminate in the unschedulable check above
+	if err := s.Schedule(ctx, pod); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to schedule, %s", err.Error())
+	}
+	return reconcile.Result{Requeue: true}, nil
+}
+
+func (s *Scheduler) Schedule(ctx context.Context, pod *v1.Pod) error {
+	var provisioner *Provisioner
+	var errs error
+	for _, candidate := range s.controller.List(ctx) {
+		if err := candidate.Spec.Supports(pod); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tried provisioner/%s: %w", candidate.Name, err))
+		} else {
+			provisioner = candidate
+			break
+		}
+	}
+	if provisioner == nil {
+		err := fmt.Errorf("matched 0/%d provisioners", len(multierr.Errors(errs)))
+		if errs != nil {
+			err = fmt.Errorf("%s, %w", err, errs)
+		}
+		return err
+	}
+	logging.FromContext(ctx).Debugf("Scheduling pod to provisioner %q", provisioner.Name)
+	if err := provisioner.Enqueue(ctx, pod); err != nil {
```

Comment:
I think this line might still be dangerous (even after the change to use the `cancelFunc` as `Stop`) because it could write to the `pods` channel after `close()` was called successful (from a different goroutine) on the said channel.

IIUC, this call to Enqueue is happening on a different goroutine from the one the closes the channel. Specifically:

```go
// Enqueue a pod to the provisioner and block until it's processed
func (p *Provisioner) Enqueue(ctx context.Context, pod *v1.Pod) error {
	p.pods <- pod // <<<===scehduler goroutine?===<<<
	return <-p.results
}
```

 could could happen after this one:

```go
func (p *Provisioner) Start(ctx context.Context) {
	go func() {
		logging.FromContext(ctx).Info("Starting provisioner")
		for {
			select {
			case <-ctx.Done():
				logging.FromContext(ctx).Info("Stopping provisioner")
				close(p.pods) // <<<=== different goroutine from scheduler ===<<<
				close(p.results)
				return
			default:
				if err := p.Provision(ctx); err != nil {
					logging.FromContext(ctx).Errorf("Provisioning failed, %s", err.Error())
				}
			}
		}
	}()
}
```

According to https://go101.org/article/channel-closing.html one should normally refrain from closing a channel from the receiving side, so one possible solution would be to have the Provisioner *not* be the thing that closes the `pods` channel. But I haven't thought it through that much.

Reply:
Trying out a solution w/ context.Done() and select statements.